### ISSUE TITLE
No results found... - for any query

### ISSUE THREAD

**Issue by GregPK** [gautamkrishnar/socli#163]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2410%20Funded-%2300A156.svg" />](https://issuehunt.io/r/gautamkrishnar/socli/issues/163)

Tried a bunch of queries but even the reference one does not seem to work:

```
gregpk[10:58] ~ ☞ socli for loop in python syntax
No results found...
gregpk[11:01] ~ ☞ socli -iq while loop in python
No results found...
```

Using `socli-3.9` installed via pip.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $10.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($10.00)

### Submitted pull Requests
- [#166 Fix inconsistent google search](https://issuehunt.io/r/gautamkrishnar/socli/pull/166)
- [#176 class name change in bs4](https://issuehunt.io/r/gautamkrishnar/socli/pull/176)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/gautamkrishnar/socli/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by isaac-philip** [comment]
Is this open ?

**Comment by gautamkrishnar** [comment]
@isaac-philip can you please look into it and verify whether the issue exists. If it is please comment on this issue. you can start working on it if you are interested.

**Comment by isaac-philip** [comment]
Yes seems to be, I am also not getting any response for same set of inputs mentioned in issue.
I will take this up, thanks!

**Comment by gautamkrishnar** [comment]
@isaac-philip please make sure that you use https://github.com/gautamkrishnar/socli/tree/refactoring branch

**Comment by isaac-philip** [comment]
> @isaac-philip please make sure that you use https://github.com/gautamkrishnar/socli/tree/refactoring branch

noted

**Comment by isaac-philip** [comment]
getting error in having same module name - socli under package - socli
clashing imports ?

Bunch of errors I could get in Python-3.6  are,

```bash
Python 3.6.2 (v3.6.2:5fd33b5926, Jul 16 2017, 20:11:06)
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import socli.user
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File " /Development/github/socli/socli/user.py", line 8, in <module>
    import socli.printer as pr
  File " /Development/github/socli/socli/printer.py", line 14, in <module>
    from socli import search as search, tui as tui
  File " /Development/github/socli/socli/search.py", line 15, in <module>
    import socli.tui
  File " /Development/github/socli/socli/tui.py", line 9, in <module>
    import socli.printer as pr
AttributeError: module 'socli' has no attribute 'printer'
```

Other similar errors depending on which package i would try to import or run.

would like to move the entire contents of socli into `__main__` and no issues would be there.
Then can just delete that module - socli.py

Let me know if thats ok and I am bit confused how things were executing till now?

Maybe I am doing something wrong ?

**Comment by gautamkrishnar** [comment]
@isaac-philip dont use python sheel use python interpreter.
you can run socli by the following command if you are in the project folder:
```
python -m socli
```
You can also try in in the shell by appending the path manually. Please see #162 by @anirudnits

**Comment by isaac-philip** [comment]
Ok sure, yes i didn't wanna add path manually.

I can execute locally as you said and will go ahead without any path appending.

**Comment by isaac-philip** [comment]
just to give an update, working on it, should expect a commit by tomorrow.

**Comment by gautamkrishnar** [comment]
@isaac-philip thanks a lot 👍 please dont forget to submit it on issuehunt https://issuehunt.io/r/gautamkrishnar/socli/issues/163

**Comment by isaac-philip** [comment]
Hey @gautamkrishnar , please checkout and run my committed code.

I have commiteed for direction since the results are random.

What happens is that after a fresh hit to the api, after few hits the result received is not as per our expectation.

Two searches are tuned to the expected html (in the commit - _html_result_expected.html_)

1. the number of question/answers - i.e. div: class_=g
2. the question/desc within each of the above div

The question/desc I could improve and that seems to be a definite part of the future commit.

I looked to utilize retries and cookies but both don't seem that much of a difference to me.

Using cookies can be beneficial but my code debugs with retries does not yield any difference. 

Its like when the google search api decides to give out the other set of html it will give it for a period of say 3-4 minutes till it reverts back to our expected html.

The retries placed in search are just for debugginng now since it doesn't yield any differennt results and only makes Google search complain that we are sending many calls/traffic.

Please advise on the direction that I should employ after you checkout and have a run to better understand my view.

**Comment by isaac-philip** [comment]
Hello @gautamkrishnar, just messaging to know if you had a look, no worries can wait.

**Comment by langenhagen** [comment]
I am also unable to retrieve results via socli in most cases.
I installed `socli` via `pipsi` the pip script installer.
Notably, `socli` seems to work when I set the flag `-s`.

My command line history is:
```
pipsi install socli
socli for loop in python  # no output
socli -q for loop in python  # No results found...
socli -q 'for loop in python'  # see above
socli -iq 'cat'  # s.a.
socli -q window.open  # s.a.
socli -iq html error 404  # s.a.
socli -iq 'for loop'  # s.a.
socli -iq 'for-loop'  # 1 result
socli -isq 'for loop'  # 10 results
```

**Comment by charliergi** [comment]
Same results here, I can give a try to fix this issue if it's not solved yet. I've tried to install socli with pip, with the master branch and the refactoring branch, and they have all the same results.

**Comment by timsmaluk** [comment]
Look like I've identified the problem. After some debugging the problem lies in this part:
```python
def get_questions_for_query_google(query, count=10):
    """
    Fetch questions for a query using Google search.
    Returned question urls are URLS to SO homepage.
    At most 10 questions are returned. (Can be altered by passing count)
    :param query: User-entered query string
    :return: list of [ (question_text, question_description, question_url) ]
    """
    i = 0
    questions = []
    randomheaders()
    search_results = requests.get(google_search_url + query, headers=header)
    captchacheck(search_results.url)
    soup = BeautifulSoup(search_results.text, 'html.parser')
    try:
        soup.find_all("div", class_="g")[0]  # For explicitly raising exception
    except IndexError:
        print_warning("No results found...")
        sys.exit(0)
    for result in soup.find_all("div", class_="g"):
        if i == count:
            break
        try:
            question_title = result.find("h3", class_="r").get_text()[:-17]
            question_desc = result.find("span", class_="st").get_text()
            if question_desc=="": # For avoiding instant answers
                raise NameError #Explicit raising
            question_url = result.find("a").get("href") #Retrieves the Stack Overflow link
            question_url = fixGoogleURL(question_url)

            if question_url is None:
                i = i-1
                continue

            questions.append([question_title, question_desc, question_url])
            i += 1
        except NameError:
            continue
        except AttributeError:
            continue

    #Check if there are any valid question posts
    if not questions:
        print_warning("No results found...")
        sys.exit(0)
    return questions
```
Specifically here:
```python
try:
        soup.find_all("div", class_="g")[0]  # For explicitly raising exception
    except IndexError:
        print_warning("No results found...")
        sys.exit(0)
```
Becuase `class_="g"` does not exit. None of the `class_=` tags exist. So the code executes the `print_warning` message and then exits. 

I'm not sure if theres any easy fix other than rewriting this function and the other scraping functions. Also, it would be nice if there were some unittests in this code. Maybe we can open up a separate PR.

**Comment by timsmaluk** [comment]
Update: I am now getting an output for the commands with -q 
![Screen Shot 2019-11-27 at 11 46 40 AM](https://user-images.githubusercontent.com/41490117/69755209-95e03500-110c-11ea-926c-ba48b88f116a.jpg)
![Screen Shot 2019-11-27 at 11 53 27 AM](https://user-images.githubusercontent.com/41490117/69755215-98db2580-110c-11ea-8638-afc2b3ba7221.jpg)

A couple things to fix here. First the statistics arent loading and I'll try to dive more into why today. Secondly in debug mode, if i step too fast through the code I get a sys.exit(0) with "No results found". I'm going to write some unnitests to figure out why that is.

**Comment by anirudnits** [comment]
I would like to work on the writing some unittests. I didn't see any new issue for this, so is it still open?

**Comment by schedutron** [comment]
If it's halted, can I work on rectifying this issue?

**Comment by timsmaluk** [comment]
It seems like the owner of this project no longer works on it. That being said it seems pointless to even try to fix it. There is a lot that needs to be done.

**Comment by gautamkrishnar** [comment]
@timsmaluk yes you are right..I am not getting time for this project due to my work related stuffs. I am looking for a maintainer who's interested in revamping the project.  @schedutron please let me know if you are interested.

**Comment by timsmaluk** [comment]
@gautamkrishnar I would be interested in becoming a maintainer. I've spent a good couple weeks on this project trying to understand the code, and I can offer some help in fixing it up to production ready code. I got it to work in my debugging environment, but the only problem was the user agents. It had about an 80% chance of working even with the new agents I implemented

**Comment by gautamkrishnar** [comment]
Hi @timsmaluk really happy to know that you're interested.  Just added you as a collaborator. Please accept the invitation by clicking on the link: https://github.com/gautamkrishnar/socli/invitations

Please note that the current development is done on the branch  https://github.com/gautamkrishnar/socli/tree/refactoring

Feel free to continue it. Please note that there are some pending pull requests to merge. As a maintainer you are free to accept or reject any of it if you think that it does match the project's roadmap. As of now the master branch is protected  from force pushing. Please open a pr from refactoring to the master branch once everything is complete. Make sure that you merge the develop branch to it too.

As of now the main goal would be making the project working again. Please let me know if this PR: https://github.com/gautamkrishnar/socli/pull/166 qualifies for the reward after trying it out. 

Feel free to ping me on: https://gitter.im/socli-community/Lobby.  Really happy to have you on board.  Thanks again for your interest 😄

**Comment by tranchikhang** [comment]
May I ask if this issue is still open?
I looked into the source code. Looks like depend on the User agent, the response HTML structure will be different.
For example, if I set User agent as (from PR: [#166](https://github.com/gautamkrishnar/socli/pull/166/commits/1e95e591a542dffc642143245246f59e6c300c91)):
`Opera/9.80 (X11; Linux x86_64; U; en-GB) Presto/2.2.15 Version/10.01`
Then the result would be:
![no_class](https://user-images.githubusercontent.com/16659747/72401093-74a3d880-378e-11ea-9120-a2d575545dc9.PNG)

As you can see, no "g" class.
So first I think we need to update the User agent file. Since the HTML structure could change, we should only include User agent that fits with the current logic.
I modified a little bit and managed to get the result:
![result_ok](https://user-images.githubusercontent.com/16659747/72401297-f4ca3e00-378e-11ea-8ea0-5f12665a927a.PNG)

**Comment by gautamkrishnar** [comment]
Thanks for finding this @tranchikhang can you open a PR so that @timsmaluk can look into it

**Comment by tranchikhang** [comment]
So I updated the code, but haven't opened a PR yet. I managed to get the data, but couldn't display the result:
![fix](https://user-images.githubusercontent.com/16659747/72435136-beb5aa00-37e0-11ea-998c-cd23233d3ed2.png)
I cannot go past this line, any ideas?
```python
tui.MAIN_LOOP = tui.EditedMainLoop(tui.question_post, printer.palette)
```

**Comment by timsmaluk** [comment]
@tranchikhang  I'm not sure what the error is since I don't have that line in my code. I had it working at one point in the debugger but I cant replicate that now. Once I figure out what wrong in my own code I can offer more help.

**Comment by tranchikhang** [comment]
@timsmaluk 
It's from here.
https://github.com/gautamkrishnar/socli/blob/141ead78b6b409981940dc15b39a1d4f4ec74636/socli/printer.py#L216
Turn out it's just a small typo, should be:
```python
tui.MAIN_LOOP = tui.EditedMainLoop(tui.question_post, palette)
```
But then I got:
```
name 'fcntl' is not defined
```
probably because [urwind cannot run on Windows](https://stackoverflow.com/questions/54357073/nameerror-name-fcntl-is-not-defined-when-using-urwid-on-windows)? I will try to check with a Linux machine later.

**Comment by timsmaluk** [comment]
@tranchikhang we should compare our code

**Comment by tranchikhang** [comment]
@timsmaluk I opened PR #174, no problem so far but I'm still new to this project so may be it needs more testing. Also I have to use WSL since urwind cannot run on Windows. Probably we should open another issue for Windows compatibility?

**Comment by sanilborkar** [comment]
So I was looking into this issue. Apparently, it seems like the ```div``` tags in the Google result webpage are rendered using JavaScript. ```requests``` package cannot deal with JavaScript rendering but its sibling ```requests_html``` can.

I tried my hands on working with that but it seems like a temporary solution, and have raised a **PR #175** [Issue #163 - Fixed Google Search not returning results](https://github.com/gautamkrishnar/socli/pull/175) for the same.

However, if I make a lot of requests, Google may end up getting my IP banned for web scraping.

---
#### Proposed Solutions
* **Rotate IP addresses** too - as @timsmaluk mentioned in Issue #170 , we can get more proxies to help us with, maybe?
* **Use Google Search API** - Of course, this is paid. This is what is mentioned on their website.
> Custom Search JSON API provides 100 search queries per day for free. If you need more, you may sign up for billing in the API Console. Additional requests cost $5 per 1000 queries, up to 10k queries per day.

> If you need more than 10k queries per day and your Custom Search Engine searches 10 sites or fewer, you may be interested in the Custom Search Site Restricted JSON API, which does not have a daily query limit.

**Comment by timsmaluk** [comment]
Okay after some time off, I am back to work on this project full time. I just got a job so I wont be avialable during the day, but I plan to put in good hours after work and no the weekend. I plan to check over all the PR's and get this project working again. Thank you for being patient

**Comment by tranchikhang** [comment]
I took a look at @sanilborkar PR #175 . I don't think the problem is JavaScript rendering. Because:
- ```request``` can fetch HTML rendered from JS fine, you can see after fetching HTML, the logic is the same: looking for ```div``` with ```class="g"```
- I also check the requests-html package, inside it has ```import requests```, and the constructor is just calling back to ```requests.Session```. The difference is we don't have to deal with User Agent if we use requests-html, since by default, it has a default User Agent which is working:
https://github.com/psf/requests-html/blob/dc82965d66761eed3b92ff371b8af9c4b85cc725/requests_html.py#L27
`DEFAULT_USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/603.3.8 (KHTML, like Gecko) Version/10.1.2 Safari/603.3.8'`

If we make a request using this UA, the rendered HTML contains ```div``` with ```class="g"```, which is why it works (not because of JavaScript rendering):
![キャプチャ](https://user-images.githubusercontent.com/16659747/73815114-4d3daa00-4828-11ea-8206-b7104887c92b.PNG)

So with all of the above, I still think that using ```requests``` and the current logic is fine: send a request -> fetch html -> look for ```div``` with ```class="g"```. But depends on the UA, the structure could change, so we need to stick with working UA only, since we don't know how many structure that Google search can generate.

**Comment by tranchikhang** [comment]
@timsmaluk Any update on this issue?

**Comment by mitchins** [comment]
Still not working for me Python 3.7.3 on OSX

**Comment by manishbhatt** [comment]
Any update? it's still not working

**Comment by hedythedev** [comment]
Hi, any progress on this? It shows no results found for any query for me. I'm on Ubuntu with Python 3.6.9, using `python3 -m socli for loop in python` shows no result as well. 

Installed with `sudo pip3 install socli`

**Comment by hedythedev** [comment]
Hi @tranchikhang and everyone, I see that this issue has not been resolved, so maybe we can set `google_search` to false, so at least the example in the README, `socli for loop in python syntax` works?

**Comment by tranchikhang** [comment]
@hedythedev 
As I said above, I opened PR #174, I just checked again today, seems to be no problem except it cannot run on Windows.
But it has been 5 months and no response, and it looks like no one is working on this project anymore, so I was planning to maintain a version using this PR onward in my repo. I'm a bit busy with work so the plan was delayed, I will try to start working on it this weekend.
![2](https://user-images.githubusercontent.com/16659747/83763050-3bbfbe00-a6b3-11ea-9611-3fdb6a1515b6.gif)

**Comment by gautamkrishnar** [comment]
@hedythedev sorry for the delayed response I am unable to maintain this project due to my work. If you are interested, I can add you as a maintainer so that you can merge that pr yourself. Please let me know if you are interested.

**Comment by gautamkrishnar** [comment]
Looks like most of the issues are caused due to some recent changes in stackoverflow and google html structure.

**Comment by hedythedev** [comment]
> @hedythedev sorry for the delayed response I am unable to maintain this project due to my work. If you are interested, I can add you as a maintainer so that you can merge that pr yourself. Please let me know if you are interested.

@gautamkrishnar 

Sure! I would love to become a maintainer. I will take some time to look at the PRs now 😀

**Comment by gautamkrishnar** [comment]
@hedythedev Thanks a lot. Just added you https://github.com/gautamkrishnar/socli/invitations

**Comment by hedythedev** [comment]
@gautamkrishnar 

I have taken a look at all 4 PRs regarding this issue, and I think #174 from @tranchikhang works fine except that:

1. Some characters aren't showing properly when passed `-iq` on WSL (see the comment from tranchikhang above)
> As I said above, I opened PR #174, I just checked again today, seems to be no problem except it cannot run on Windows.
2. When used normally like this `socli for loop in Python syntax`, it showed `could not load statistics` in the question section for some reason?

Also I haven't tested it on Git Bash for Windows and CMD/powershell yet, because there were some problems with my windows environment yesterday. But I can try it again today. So far I've tested it working fine on Gitpod with Linux environment.
If the problem on Windows can't be solved for the moment, maybe we can merge #174 so they work on non-windows environments?

**Comment by gautamkrishnar** [comment]
@hedythedev you can ignore support for wsl for now. Maybe we open an issue with proper info so that someone can fix it once we release a stable version.

**Comment by hedythedev** [comment]
Just merged it. I'm not quite sure about the workflow of how the refactoring branch can be merged to master and then released on PyPI, but I guess it is at least solved for non-windows environments.

**Comment by gautamkrishnar** [comment]
@hedythedev can you please test the latest code on the master. I just merged the refactoring branch with the master. Some tests are failing for python 2: https://travis-ci.org/github/gautamkrishnar/socli/builds/695657072 

We need to investigate that. If it's not fixable we may need to remove python 2 from the list of the supported versions.

You can always create a new release to PyPI by creating a new version from the repo: https://github.com/gautamkrishnar/socli/releases

**Comment by gautamkrishnar** [comment]
@hedythedev removing support for 2.7 would be better since it is reaching its EOL soon. I will update the code to make sure that the project uses python 3+.

**Comment by gautamkrishnar** [comment]
This issue is fixed. Will be released on version 4.0

**Comment by hedythedev** [comment]
> You can always create a new release to PyPI by creating a new version from the repo: https://github.com/gautamkrishnar/socli/releases

Got it thank you. Sorry for the late response because I think our time zones are different. I have commented on #177 that 

> We should update the [requirements section of the README](https://github.com/gautamkrishnar/socli#requirements) to `python 3.5 or higher`
